failed to commit journal after retried 1 times.

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Expected behavior (Required)
start fe successfully.
### Real behavior (Required)

```
2022-06-21 12:44:37,218 WARN (replayer|74) [BDBJournalCursor.next():148] Catch an exception when get next JournalEntity. key:9
4775
com.sleepycat.je.LockTimeoutException: (JE 7.3.7) Lock expired. Locker 845225335 -1_replayer_ReplicaThreadLocker: waited for l
ock on database=77124 LockAddr:2085329640 LSN=0xa/0x71d943 type=READ grant=WAIT_NEW timeoutMillis=1000 startTime=1655786676218
 endTime=1655786677218
Owners: [<LockInfo locker="<ReplayTxn id="-92621">399657264 -92621_ReplayThread_ReplayTxn" type="WRITE"/>]
Waiters: [<LockInfo locker="845225335 -1_replayer_ReplicaThreadLocker" type="READ"/>]

        at com.sleepycat.je.txn.LockManager.makeTimeoutException(LockManager.java:1117) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.txn.LockManager.waitForLock(LockManager.java:606) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.txn.LockManager.lock(LockManager.java:345) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.txn.BasicLocker.lockInternal(BasicLocker.java:124) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.txn.ReplicaThreadLocker.lockInternal(ReplicaThreadLocker.java:63) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.txn.Locker.lock(Locker.java:499) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.CursorImpl.lockLN(CursorImpl.java:3585) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.CursorImpl.lockLN(CursorImpl.java:3316) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.CursorImpl.lockLNAndCheckDefunct(CursorImpl.java:2138) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.CursorImpl.searchExact(CursorImpl.java:1950) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Cursor.searchExact(Cursor.java:4194) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Cursor.searchNoDups(Cursor.java:4055) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Cursor.search(Cursor.java:3857) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Cursor.getInternal(Cursor.java:1284) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Database.get(Database.java:1271) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Database.get(Database.java:1330) ~[je-7.3.7.jar:7.3.7]
        at com.starrocks.journal.bdbje.CloseSafeDatabase.get(CloseSafeDatabase.java:47) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:108) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1690) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$4.runOneCycle(GlobalStateMgr.java:1487) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
```
```
2022-06-21 14:13:07,112 WARN (starrocks-mysql-nio-pool-25|207) [FrontendServiceProxy.call():29] call frontend thrift rpc failed, addr: TNetworkAddress(hostname:172.26.92.202, port:9822), retried: 0
org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
        at org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:159) ~[libthrift-0.13.0.jar:0.13.0]
        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:73) ~[libthrift-0.13.0.jar:0.13.0]
        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:62) ~[libthrift-0.13.0.jar:0.13.0]
        at com.starrocks.thrift.FrontendService$Client.send_forward(FrontendService.java:438) ~[starrocks-fe.jar:?]
        at com.starrocks.thrift.FrontendService$Client.forward(FrontendService.java:430) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.MasterOpExecutor.lambda$forward$0(MasterOpExecutor.java:138) ~[starrocks-fe.jar:?]
        at com.starrocks.rpc.FrontendServiceProxy.call(FrontendServiceProxy.java:25) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.MasterOpExecutor.forward(MasterOpExecutor.java:135) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.MasterOpExecutor.execute(MasterOpExecutor.java:73) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.forwardToMaster(StmtExecutor.java:561) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:351) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:282) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:430) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:666) ~[starrocks-fe.jar:?]
        at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
Caused by: java.net.SocketException: Broken pipe (Write failed)
        at java.net.SocketOutputStream.socketWrite0(Native Method) ~[?:1.8.0_252]
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111) ~[?:1.8.0_252]
        at java.net.SocketOutputStream.write(SocketOutputStream.java:155) ~[?:1.8.0_252]
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82) ~[?:1.8.0_252]
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140) ~[?:1.8.0_252]
        at org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:157) ~[libthrift-0.13.0.jar:0.13.0]
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+----------------------+
| current_version()    |
+----------------------+
| MAIN-RELEASE 0eaab02 |
+----------------------+
1 row in set (0.00 sec)
```

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -539,7 +539,6 @@ public class BDBJEJournal implements Journal {
                 LOG.error(errMsg, e);
                 exception = new JournalException(errMsg);
                 exception.initCause(e);
-
             }
         }
         // failed after retried

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -508,41 +508,45 @@ public class BDBJEJournal implements Journal {
         }
 
         JournalException exception = null;
-        for (int i = 0; i < RETRY_TIME; i++) {
-            // retry cleanups
-            if (i != 0) {
-                Thread.sleep(SLEEP_INTERVAL_SEC * 1000);
+        try {
+            for (int i = 0; i < RETRY_TIME; i++) {
+                // retry cleanups
+                if (i != 0) {
+                    Thread.sleep(SLEEP_INTERVAL_SEC * 1000);
 
-                if (currentTrasaction == null || !currentTrasaction.isValid()) {
-                    try {
-                        rebuildCurrentTransaction();
-                    } catch (JournalException e) {
-                        // failed to rebuild txn, will continune to next attempt
-                        LOG.warn("failed to commit journal after retried {} times! failed to rebuild txn",
-                                i + 1, e);
-                        currentTrasaction = null;
-                        exception = e;
-                        continue;
+                    if (currentTrasaction == null || !currentTrasaction.isValid()) {
+                        try {
+                            rebuildCurrentTransaction();
+                        } catch (JournalException e) {
+                            // failed to rebuild txn, will continune to next attempt
+                            LOG.warn("failed to commit journal after retried {} times! failed to rebuild txn",
+                                    i + 1, e);
+                            currentTrasaction = null;
+                            exception = e;
+                            continue;
+                        }
                     }
+                } // if i != 0
+
+                // commit
+                try {
+                    currentTrasaction.commit();
+                    return;
+                } catch (DatabaseException e) {
+                    String errMsg = String.format("failed to commit journal after retried %d times! txn[%s] db[%s]",
+                            i + 1, currentTrasaction, currentJournalDB);
+                    LOG.error(errMsg, e);
+                    exception = new JournalException(errMsg);
+                    exception.initCause(e);
                 }
             }
-
-            // commit
-            try {
-                currentTrasaction.commit();
-                currentTrasaction = null;
-                uncommitedDatas.clear();
-                return;
-            } catch (DatabaseException e) {
-                String errMsg = String.format("failed to commit journal after retried %d times! txn[%s] db[%s]",
-                        i + 1, currentTrasaction, currentJournalDB);
-                LOG.error(errMsg, e);
-                exception = new JournalException(errMsg);
-                exception.initCause(e);
-            }
+            // failed after retried
+            throw exception;
+        } finally {
+            // always reset current txn
+            currentTrasaction = null;
+            uncommitedDatas.clear();
         }
-        // failed after retried
-        throw exception;
     }
 
     /**

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -46,6 +46,7 @@ import org.apache.logging.log4j.Logger;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicLong;
 
@@ -59,8 +60,8 @@ import java.util.concurrent.atomic.AtomicLong;
 public class BDBJEJournal implements Journal {
     public static final Logger LOG = LogManager.getLogger(BDBJEJournal.class);
     private static final int OUTPUT_BUFFER_INIT_SIZE = 128;
-    static final int RETRY_TIME = 3;
-    static final int SLEEP_INTERVAL_SEC = 5;
+    static int RETRY_TIME = 3;
+    static int SLEEP_INTERVAL_SEC = 5;
 
     private String environmentPath = null;
     private String selfNodeName;
@@ -69,6 +70,10 @@ public class BDBJEJournal implements Journal {
     private BDBEnvironment bdbEnvironment = null;
     private CloseSafeDatabase currentJournalDB;
     protected Transaction currentTrasaction = null;
+
+    // store uncommitted kv, used for rebuilding txn on commit fails
+    private List<Pair<DatabaseEntry, DatabaseEntry>> uncommitedDatas = new ArrayList<>();
+
     // only kept for profile test, will remove in the next PR
     @VisibleForTesting
     private AtomicLong nextJournalId = new AtomicLong(1);
@@ -472,6 +477,7 @@ public class BDBJEJournal implements Journal {
                             i + 1, status, currentJournalDB, theKey, theData));
                 }
                 // success
+                uncommitedDatas.add(Pair.create(theKey, theData));
                 return;
             } catch (DatabaseException e) {
                 String errMsg = String.format(
@@ -492,6 +498,8 @@ public class BDBJEJournal implements Journal {
     /**
      * persist current batch
      * for bdb: commit current transaction
+     * notice that if commit fail, the transaction may not be valid.
+     * we should rebuild the transaction and retry.
      */
     @Override
     public void batchWriteCommit() throws InterruptedException, JournalException {
@@ -501,14 +509,29 @@ public class BDBJEJournal implements Journal {
 
         JournalException exception = null;
         for (int i = 0; i < RETRY_TIME; i++) {
-            try {
-                // sleep before retry
-                if (i != 0) {
-                    Thread.sleep(SLEEP_INTERVAL_SEC * 1000);
-                }
+            // retry cleanups
+            if (i != 0) {
+                Thread.sleep(SLEEP_INTERVAL_SEC * 1000);
 
+                if (currentTrasaction == null || !currentTrasaction.isValid()) {
+                    try {
+                        rebuildCurrentTransaction();
+                    } catch (JournalException e) {
+                        // failed to rebuild txn, will continune to next attempt
+                        LOG.warn("failed to commit journal after retried {} times! failed to rebuild txn",
+                                i + 1, e);
+                        currentTrasaction = null;
+                        exception = e;
+                        continue;
+                    }
+                }
+            }
+
+            // commit
+            try {
                 currentTrasaction.commit();
                 currentTrasaction = null;
+                uncommitedDatas.clear();
                 return;
             } catch (DatabaseException e) {
                 String errMsg = String.format("failed to commit journal after retried %d times! txn[%s] db[%s]",
@@ -516,10 +539,46 @@ public class BDBJEJournal implements Journal {
                 LOG.error(errMsg, e);
                 exception = new JournalException(errMsg);
                 exception.initCause(e);
+
             }
         }
         // failed after retried
         throw exception;
+    }
+
+    /**
+     * txn can be invalid if commit fails on exception
+     * in this case, we rebuild the current transaction with `uncommitedDatas`
+     * there's no need to retry while we were rebuilding since we have retried outside this function
+     */
+    private void rebuildCurrentTransaction() throws JournalException {
+        LOG.warn("transaction is invalid, rebuild the txn with {} kvs", uncommitedDatas.size());
+
+        try {
+            //  begin transaction
+            currentTrasaction = currentJournalDB.getDb().getEnvironment().beginTransaction(
+                    null, bdbEnvironment.getTxnConfig());
+            // append
+            for (Pair<DatabaseEntry, DatabaseEntry> kvPair : uncommitedDatas) {
+                DatabaseEntry theKey = kvPair.first;
+                DatabaseEntry theData = kvPair.second;
+                OperationStatus status = currentJournalDB.put(currentTrasaction, theKey, theData);
+                if (status != OperationStatus.SUCCESS) {
+                    String msg = String.format(
+                            "failed to append journal! status[%s] db[%s] key[%s] data[%s]",
+                            status, currentJournalDB, theKey, theData);
+                    LOG.warn(msg);
+                    throw new JournalException(msg);
+                }
+            }
+            LOG.info("rebuild txn succeed. new txn {}", currentTrasaction);
+        } catch (DatabaseException e) {
+            String errMsg = String.format("failed to rebuild txn! txn[%s] db[%s]", currentTrasaction, currentJournalDB);
+            LOG.error(errMsg, e);
+            JournalException exception = new JournalException(errMsg);
+            exception.initCause(e);
+            throw exception;
+        }
     }
 
     /**

--- a/fe/fe-core/src/test/java/com/starrocks/journal/bdbje/BDBJEJournalTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/journal/bdbje/BDBJEJournalTest.java
@@ -330,8 +330,9 @@ public class BDBJEJournalTest {
         journal.batchWriteAppend(1, buffer);
     }
 
+
     @Test
-    public void testBatchWriteCommitRetry(
+    public void testBatchWriteCommitRetrySuccess(
             @Mocked CloseSafeDatabase database,
             @Mocked BDBEnvironment environment,
             @Mocked Database rawDatabase,
@@ -394,6 +395,77 @@ public class BDBJEJournalTest {
 
         journal.batchWriteBegin();
         journal.batchWriteAppend(1, buffer);
+        journal.batchWriteCommit();
+    }
+
+    // retry 1: commit fails
+    // retry 2: begin txn with exception
+    // retry 3: put return error
+    @Test(expected = JournalException.class)
+    public void commitFailsRebuildFails(
+            @Mocked CloseSafeDatabase database,
+            @Mocked BDBEnvironment environment,
+            @Mocked Database rawDatabase,
+            @Mocked Environment rawEnvironment,
+            @Mocked Transaction txn) throws Exception {
+        BDBJEJournal journal = new BDBJEJournal(environment, database);
+        String data = "petals on a wet black bough";
+        DataOutputBuffer buffer = new DataOutputBuffer();
+        Text.writeString(buffer, data);
+
+        // write failed and retry, success
+        new Expectations(database) {
+            {
+                database.getDb();
+                minTimes = 0;
+                result = rawDatabase;
+
+                database.put((Transaction) any, (DatabaseEntry) any, (DatabaseEntry) any);
+                times = 4;  // append & rebuild txn
+                result = OperationStatus.SUCCESS;
+                result = OperationStatus.SUCCESS;
+                result = OperationStatus.SUCCESS;
+                result = OperationStatus.KEYEMPTY;
+            }
+        };
+
+        new Expectations(rawDatabase) {
+            {
+                rawDatabase.getEnvironment();
+                minTimes = 0;
+                result = rawEnvironment;
+
+                rawDatabase.getDatabaseName();
+                minTimes = 0;
+                result = "fakeDb";
+            }
+        };
+
+        new Expectations(rawEnvironment) {
+            {
+                rawEnvironment.beginTransaction(null, (TransactionConfig)any);
+                times = 3;  // append & rebuild txn
+                result = txn;
+                result = new DatabaseNotFoundException("mock mock: begin txn");
+                result = txn;
+            }
+        };
+
+        new Expectations(txn) {
+            {
+                txn.commit();
+                times = 1;
+                result = new DatabaseNotFoundException("mock mock: commit");
+
+                txn.isValid();
+                times = 1;
+                result = false;
+            }
+        };
+
+        journal.batchWriteBegin();
+        journal.batchWriteAppend(1, buffer);
+        journal.batchWriteAppend(2, buffer);
         journal.batchWriteCommit();
     }
 

--- a/fe/fe-core/src/test/java/com/starrocks/journal/bdbje/BDBJEJournalTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/journal/bdbje/BDBJEJournalTest.java
@@ -47,6 +47,8 @@ public class BDBJEJournalTest {
 
     @Before
     public void init() throws Exception{
+        BDBJEJournal.RETRY_TIME = 3;
+        BDBJEJournal.SLEEP_INTERVAL_SEC = 0;
         // avoid checkpoint
         restoredCheckpointIntervalSecond = FeConstants.checkpoint_interval_second;
         FeConstants.checkpoint_interval_second = 24 * 60 * 60;
@@ -260,13 +262,6 @@ public class BDBJEJournalTest {
         DataOutputBuffer buffer = new DataOutputBuffer();
         Text.writeString(buffer, data);
 
-        new Expectations() {
-            {
-                Thread.sleep(anyLong);
-                times = 1;
-            }
-        };
-
         new Expectations(rawDatabase) {
             {
                 rawDatabase.getEnvironment();
@@ -297,12 +292,6 @@ public class BDBJEJournalTest {
         DataOutputBuffer buffer = new DataOutputBuffer();
         Text.writeString(buffer, data);
 
-        new Expectations() {
-            {
-                Thread.sleep(anyLong);
-                times = 2;
-            }
-        };
         new Expectations(database) {
             {
                 database.getDb();
@@ -354,13 +343,6 @@ public class BDBJEJournalTest {
         Text.writeString(buffer, data);
 
         // write failed and retry, success
-        new Expectations() {
-            {
-                Thread.sleep(anyLong);
-                times = 1;
-            }
-        };
-
         new Expectations(database) {
             {
                 database.getDb();
@@ -368,7 +350,7 @@ public class BDBJEJournalTest {
                 result = rawDatabase;
 
                 database.put((Transaction) any, (DatabaseEntry) any, (DatabaseEntry) any);
-                times = 1;
+                times = 2;  // append & rebuild txn
                 result = OperationStatus.SUCCESS;
             }
         };
@@ -388,16 +370,25 @@ public class BDBJEJournalTest {
         new Expectations(rawEnvironment) {
             {
                 rawEnvironment.beginTransaction(null, (TransactionConfig)any);
-                times = 1;
+                times = 2;  // append & rebuild txn
             }
         };
 
+        // commit fails 2 times
+        // the first txn is valid, can continue commit
+        // the second time txn is invalid, we have to rebuild txn
         new Expectations(txn) {
             {
                 txn.commit();
-                times = 2;
+                times = 3;
+                result = new DatabaseNotFoundException("mock mock");
                 result = new DatabaseNotFoundException("mock mock");
                 result = null;
+
+                txn.isValid();
+                times = 2;
+                result = true;
+                result = false;
             }
         };
 

